debian/matrix-synapse-py3.preinst

```diff
@@ -1,31 +0,0 @@
-#!/bin/sh -e
-
-# Attempt to undo some of the braindamage caused by
-# https://github.com/matrix-org/package-synapse-debian/issues/18.
-#
-# Due to reasons [1], the old python2 matrix-synapse package will not stop the
-# service when the package is uninstalled. Our maintainer scripts will do the
-# right thing in terms of ensuring the service is enabled and unmasked, but
-# then do a `systemctl start matrix-synapse`, which of course does nothing -
-# leaving the old (py2) service running.
-#
-# There should normally be no reason for the service to be running during our
-# preinst, so we assume that if it *is* running, it's due to that situation,
-# and stop it.
-#
-# [1] dh_systemd_start doesn't do anything because it sees that there is an
-#     init.d script with the same name, so leaves it to dh_installinit.
-#
-#     dh_installinit doesn't do anything because somebody gave it a --no-start
-#     for unknown reasons.
-
-if [ -x /bin/systemctl ]; then
-    if /bin/systemctl --quiet is-active -- matrix-synapse; then
-        echo >&2 "stopping existing matrix-synapse service"
-        /bin/systemctl stop matrix-synapse || true
-    fi
-fi
```

Comment:
I believe this is still necessary. Can you explain why you don't think it is?

Reply:
I mean, I guess it's reasonable to stop supporting people upgrading from a four-year-old version at some point. It's just not obvious to me what that has to do with changing the debhelper version.

Reply:
Granted, it is not strictly necessary for upgrading the debhelper version, but it fits thematically. The reasoning references only debhelper modules that are no longer used (at least with this PR) and this does has the added benefit of making updates more seemless, because downtime is reduced from "stop in preinst, start in postinst" to "restart in postinst".

If you wish, I can drop the commit from this PR and open another one, but I the value of more seemless updates (and less moving parts not managed by Debian's machinery itself) outweighs keeping support around for updating a four-year old synapse versions on an EOL Python version (I'd also be hesitant trying to do an upgrade over this many versions in the first place).

Reply:
> The reasoning references only debhelper modules that are no longer used (at least with this PR)

well yes, it's working around a bug in a previous version of the package. Of course that version used old debhelper modules.

> is does has the added benefit of making updates more seemless, because downtime is reduced from "stop in preinst, start in postinst"

Yeah... it seems `dh_systemd_start`'s behaviour changed here. The logic in this `preinst` assumed that normally the service would be stopped in the preinst and started in the postinst; that changed in debhelper-compat 10. Replacing all the python files while Synapse is still running sounds... brave; but evidently it's been working fine since https://github.com/matrix-org/synapse/pull/10429.

So yes, I agree it's time to get rid of this. I'd probably have preferred it to be in a separate PR, but now that we're here, I guess it might as well stay.

Reply:
> Replacing all the python files while Synapse is still running sounds... brave

Yeah, I definitely know what you think, but I assume that - at least as long as cpython's imports are not lazy and there is no memory pressure -, this is asymptotically safe, because the imports will have already loaded been loaded into memory, but that's more my handwavy thinking why this works.